reconciler/leader.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconciler
+
+import (
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// Bucket is an opaque type used to scope leadership.
+type Bucket interface {
+	// Name returns a string representing this bucket, which uniquely
+	// identifies the bucket and is suitable for use as a resource lock name.
+	Name() string
+
+	// Has determines whether this Bucket contains a particular key.
+	Has(key types.NamespacedName) bool
+}
+
+// UniversalBucket returns a Bucket that "Has()" all keys.
+func UniversalBucket() Bucket {
+	return &bucket{}
+}
+
+type bucket struct{}
+
+var _ Bucket = (*bucket)(nil)
+
+// Name implements Bucket
+func (b *bucket) Name() string {
+	return ""
+}
+
+// Has implements Bucket
+func (b *bucket) Has(nn types.NamespacedName) bool {
+	return true
+}
+
+// LeaderAware is implemented by Reconcilers that are aware of their leader status.
+type LeaderAware interface {
+	// Promote is called when we become the leader of a given Bucket.  It must be
+	// supplied with an enqueue function through which a Bucket resync may be triggered.
+	Promote(b Bucket, enq func(Bucket, types.NamespacedName))
```

Comment:
I think we should allow the `Promote` callback to signal if there were any errors. This should allow the leader election runners to cancel the lease. Looking at the changes in the `genreconciler` this would happen if we failed to list all objects in order to re-sync.

```suggestion
	Promote(b Bucket, enq func(Bucket, types.NamespacedName)) error
```

Reply:
The leader election package doesn't provide us with any facility to abdicate, and technically the only way for the lister to return an error is for `meta.Accessor` to fail, which seems virtually impossible.

Reply:
> The leader election package doesn't provide us with any facility to abdicate

Cancelling the context passed to `LeaderElector` allows you to release the lock.

Reply:
sunk up on slack, this returns an error, but we still log fatal for now (w/ TODO)